zkevm-circuits/src/bytecode_circuit/circuit.rs

```diff
@@ -0,0 +1,1041 @@
+
+        // Subtract the unusable rows from the size
+        assert!(size > self.minimum_rows);
+        let last_row_offset = size - self.minimum_rows + 1;
+
+        trace!(
+            "size: {}, minimum_rows: {}, last_row_offset:{}",
+            size,
+            self.minimum_rows,
+            last_row_offset
+        );
+
+        let empty_hash = challenges.evm_word().map(|challenge| {
+            RandomLinearCombination::<F, 32>::random_linear_combine(*EMPTY_HASH_LE, challenge)
+        });
+
+        layouter.assign_region(
+            || "assign bytecode",
+            |mut region| {
+                let mut offset = 0;
+                for bytecode in witness.iter() {
+                    self.assign_bytecode(
+                        &mut region,
+                        bytecode,
+                        challenges,
+                        &push_data_left_is_zero_chip,
+                        empty_hash,
+                        &mut offset,
+                        last_row_offset,
+                        fail_fast,
+                    )?;
+                }
+
+                // Padding
+                for idx in offset..=last_row_offset {
+                    self.set_padding_row(
+                        &mut region,
+                        &push_data_left_is_zero_chip,
+                        empty_hash,
+                        idx,
+                        last_row_offset,
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn assign_bytecode(
+        &self,
+        region: &mut Region<'_, F>,
+        bytecode: &UnrolledBytecode<F>,
+        challenges: &Challenges<Value<F>>,
+        push_rindex_is_zero_chip: &IsZeroChip<F>,
+        empty_hash: Value<F>,
+        offset: &mut usize,
+        last_row_offset: usize,
+        fail_fast: bool,
+    ) -> Result<(), Error> {
+        // Run over all the bytes
+        let mut push_data_left = 0;
+        let mut next_push_data_left = 0;
+        let mut push_data_size = 0;
+        let mut value_rlc = challenges.keccak_input().map(|_| F::zero());
+        let length = F::from(bytecode.bytes.len() as u64);
+
+        for (idx, row) in bytecode.rows.iter().enumerate() {
+            if fail_fast && *offset > last_row_offset {
+                log::error!(
+                    "Bytecode Circuit: offset={} > last_row_offset={}",
+                    offset,
+                    last_row_offset
+                );
+                return Err(Error::Synthesis);
+            }
+
+            // TODO: why different code_hash for each row? Is this going to produce the same
+            // result for every row?
+            let code_hash = challenges.evm_word().map(|challenge| {
+                RandomLinearCombination::<F, 32>::random_linear_combine(
+                    row.code_hash.to_le_bytes(),
+                    challenge,
+                )
+            });
```

Comment:
If we want we I think we could group rows of the same bytecode inside `UnrolledBytecode` into a same struct, then we can cache this value.

Reply:
This method does a bytecode as a time, so I think it can be solved just moving the expression outside the loop and using the code_hash from the first row of the bytecode.

However, this make the tests bytecode_invalid_hash_data don't pass as the code_hash of rows different from the header row are ignored by the witness generation. Hence, I need to modify this tests, but in the future we should inject an evil witness to test other constrains.

Reply:
Done